add freezeElevation logic to AnimationOptions

prozessor13 opened the pull request:
When Terrain is activated all camera movements had the *freezeElevation* logic enabled. This means that the camera will stay in the same height, based on sea-level, and on the `moveend` Event the zoomlevel is recalculated for the distance between the camera and the center-coordinate in current terrain.
But this behaviour is only needed when panning the map horizontally, and when you not care about the zoomlevel. In all other cases this feature is not needed, so it is now disabled by default.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

HarelM commented on src/ui/camera.ts:
```diff
@@ -889,16 +894,17 @@ abstract class Camera extends Evented {
     }
 
-    _prepareEase(eventData: any, noMoveStart: boolean, currently: any = {}) {
+    _prepareEase(options: any, eventData: any, noMoveStart: boolean, currently: any = {}) {
```
Better to have this typed instead of `any`...

HarelM commented on src/ui/camera.ts:
```diff
@@ -112,13 +112,18 @@ export type FitBoundsOptions = FlyToOptions & {
  * @property {boolean} animate If `false`, no animation will occur.
  * @property {boolean} essential If `true`, then the animation is considered essential and will not be affected by
  *   [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion).
+ * @property {boolean} freezeElevation When doing a map-animation in terrain, and because the camera calcalates its
```
What's the default?

HarelM commented on src/ui/camera.ts:
```diff
@@ -926,14 +932,17 @@ abstract class Camera extends Evented {
     }
 
-    _afterEase(eventData?: any, easeId?: string) {
+    _afterEase(options: any, eventData?: any, easeId?: string) {
```
Better to have this typed instead of `any`...
It is true for the other types as well...

prozessor13 commented:
related to https://github.com/maplibre/maplibre-gl-js/issues/1492#issuecomment-1219729877 , and the fact that the camera would wobble up and down during flight in heavy terrain, i think this PR is only the half way to success. My thoughts for a success:

Instead of freezing elevation, do a interpolation from start-altitude to destination-altitude. The problem with that is, that the destination-altitude may change during flight, because at start the destination may has no terrain loaded, which will loading during flight. So the linear interpolation has to be recalculated in such situations.

prozessor13 commented:
@HarelM thx for your review! but i have to rewirte this PR and i will keep your comments in mind!

prozessor13 commented:
I implemented a linear interpolation from the start center-elevation to the destination center-elevation. To do this it was necessary to set the terrain object as an instance variable of `camera`.
I thought a lot about alternatives, and to continue the way that the `terrain` object is only an instance of `style`. But because neither `transform` nor `camera` do know anything about the style-object the only way to implement a communication would be via an event, like `freezeElevation`.
But because i did not like the `freezeElevation` event from the beginning on, i removed this logic and i decided to put the terrain object to `camera`.

My questions are:

* should i remove the terrain-object from style? (because via style.map.terrain it is also accessable)
* should i move the terrain-object to transform, instead of camera? (i am not shure what is the correct place of)
* leave it as it is, and strickly communicate via events. In this case callback methods as event-arguments will needed. (for me the sounds really ugly an unintuitive)

HarelM commented:
I'll look more deeply into it later on, but it might be that the terrain object is not well defined in terms of responsibility and this is why we are having a hard time placing it (probably has too many responsibilities).
Transform is a math utility, so I don't think it should be there, this hasn't changed from my perspective from the last time we talked.
I still need to better understand what's the definition of camera and to see if the terrain object should be there or not...

Thanks for pushing this forward!

HarelM commented:
I get the feeling that `Camera` should be a separate class from `Map` or that `Map` should not inherit from `Camera` as `Map` isn't a `Camera` by inheritance definition. It might be better to have camera composition and have the camera as a member of Map.
Regardless, terrain object is too bit, but I don't know how to split it.
It seems that it has both render logic - texture etc, and conversion logic (RTT, elevation, coordinates) which uses the GPU but are not render logic completely.
I don't know how to solve this though...
If they use a framebuffer in order to do some calculation, there's no reason not to be in trabsform (I think) but the minute you add texture there, you break it...
I know there are optimizations involved in these concepts and that's why it is coupled, so I'm not 100% sure how to solve this... :-(

prozessor13 commented:
I totally agree with your opinion that camera should be an instance of *Map*. And i agree that it would be good to refacture some parts of the terrain-logic. But i think this task can only come to success with someone as sparring partner! I already refactured by my own with limited success.
someone is willing to do this together? :)

HarelM commented:
Sure, I'm with you :-) I can start a branch for making camera an instance member, see that everything passes and then we can continue and look into this part.

prozessor13 commented:
Cool :D really looking forward to get some inputs!! thx!

prozessor13 commented:
another problem related to https://github.com/maplibre/maplibre-gl-js/issues/1492#issuecomment-1231238164. i think it is needed to implement preload, or at least "very high" prio loading the terrain-tile on target-coordinate. Without a known target terrain-height it is hard to calculate the correct animation.

prozessor13 commented:
I think this code is production-ready. How to deal with the changelog entry?

HarelM commented:
Phase address the code review comments I wrote.
Edit the changelog file manually and write the fix you did and the link to the PR or the issue.

HarelM commented on src/style/style.ts:
```diff
@@ -495,28 +494,21 @@ class Style extends Evented {
         if (!options) {
             if (this.terrain) this.terrain.sourceCache.destruct();
             this.terrain = null;
+            this.map.terrain = null;
```
If the terrain object is part of the map as part of this commit, can we remove it from the style then?
Feels weird to keep them in sync...

HarelM commented:
@prozessor13 I've added one last comment before this can be merged - and this is to move the terrain object to the map and not keep it in the style. If this is possible please change this and ping me so I would merge this.
